src/databasetasks.cpp

```diff
@@ -85,6 +85,13 @@ void DatabaseTasks::runTask(const DatabaseTask& task)
 
 void DatabaseTasks::flush()
 {
+	taskLock.lock();
+	flushInternal();
+	taskLock.unlock();
+}
+
+void DatabaseTasks::flushInternal()
+{
 	while (!tasks.empty()) {
 		runTask(tasks.front());
```

Comment:
Okay, you're right, I forgot the callbacks are not called directly. Still, you're performing database IO while holding a mutex, so that's a bad idea. You should use a unique_lock in this case:
```
std::unique_lock<std::mutex> guard{taskLock}
while (!tasks.empty()) {
    auto task = std::move(tasks.front())
    tasks.pop_front();
    guard.unlock();
    runTask(task);
    guard.lock();
}
```